### ISSUE TITLE
New compilation warning in blead

### ISSUE THREAD

**Issue by dur-randir** [Perl/perl5#17664]
This is a bug report for perl from sergey.aleynikov@gmail.com,
generated with the help of perlbug 1.41 running under perl 5.31.10.

[Please describe your issue here]

```
Opcode.xs: In function ‘XS_Opcode_invert_opset’:
Opcode.xs:363:25: warning: overflow in conversion from ‘int’ to ‘char’ changes value from ‘(int)*(bitmap + ((sizetype)my_cxt.x_opset_len + 18446744073709551615)) & -256’ to ‘0’ [-Woverflow]
  bitmap[opset_len-1] &= ~(0xFF << (PL_maxo & 0x07));
                         ^
```

I've spotted this warning when compiling a fresh blead and couldn't remember seeing it earlier. There're no recent changes in ext/Opcode, so it's likely introduced by some unrelated change in types.

[Please do not change anything below this line]
Flags:
category=core
severity=medium
Site configuration information for perl 5.31.10:

Configured by root at Fri Mar 13 17:15:02 MSK 2020.

Summary of my perl5 (revision 5 version 31 subversion 10) configuration:
Commit id: 0c96aa4
Platform:
osname=linux
osvers=4.19.0-8-amd64
archname=x86_64-linux
uname='linux dorothy 4.19.0-8-amd64 #1 smp debian 4.19.98-1 (2020-01-26) x86_64 gnulinux '
config_args='-de -Dusedevel -Doptimize=-O2'
hint=recommended
useposix=true
d_sigaction=define
useithreads=undef
usemultiplicity=undef
use64bitint=define
use64bitall=define
uselongdouble=undef
usemymalloc=n
default_inc_excludes_dot=define
bincompat5005=undef
Compiler:
cc='cc'
ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
optimize='-O2'
cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
ccversion=''
gccversion='8.3.0'
gccosandvers=''
intsize=4
longsize=8
ptrsize=8
doublesize=8
byteorder=12345678
doublekind=3
d_longlong=define
longlongsize=8
d_longdbl=define
longdblsize=16
longdblkind=3
ivtype='long'
ivsize=8
nvtype='double'
nvsize=8
Off_t='off_t'
lseeksize=8
alignbytes=8
prototype=define
Linker and Libraries:
ld='cc'
ldflags =' -fstack-protector-strong -L/usr/local/lib'
libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
libc=libc-2.28.so
so=so
useshrplib=false
libperl=libperl.a
gnulibc_version='2.28'
Dynamic Linking:
dlsrc=dl_dlopen.xs
dlext=so
d_dlsymun=undef
ccdlflags='-Wl,-E'
cccdlflags='-fPIC'
lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'

@inc for perl 5.31.10:
lib
/usr/local/lib/perl5/site_perl/5.31.10/x86_64-linux
/usr/local/lib/perl5/site_perl/5.31.10
/usr/local/lib/perl5/5.31.10/x86_64-linux
/usr/local/lib/perl5/5.31.10

Environment for perl 5.31.10:
HOME=/home/afl
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE=en_US.UTF-8
LC_TIME=C
LD_LIBRARY_PATH (unset)
LOGDIR (unset)
PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.30.0-dbg/bin:/opt/local/bin:/usr/texbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PERLBREW_HOME=/home/afl/.perlbrew
PERLBREW_MANPATH=/home/afl/perlbrew/perls/perl-5.30.0-dbg/man
PERLBREW_PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.30.0-dbg/bin
PERLBREW_PERL=perl-5.30.0-dbg
PERLBREW_ROOT=/home/afl/perlbrew
PERLBREW_SHELLRC_VERSION=0.88
PERLBREW_VERSION=0.88
PERL_BADLANG (unset)

**Comment by jkeenan** [comment]
> This is a bug report for perl from [sergey.aleynikov@gmail.com](mailto:sergey.aleynikov@gmail.com),
> generated with the help of perlbug 1.41 running under perl 5.31.10.
> 
> [Please describe your issue here]
> 
> ```
> Opcode.xs: In function ‘XS_Opcode_invert_opset’:
> Opcode.xs:363:25: warning: overflow in conversion from ‘int’ to ‘char’ changes value from ‘(int)*(bitmap + ((sizetype)my_cxt.x_opset_len + 18446744073709551615)) & -256’ to ‘0’ [-Woverflow]
>   bitmap[opset_len-1] &= ~(0xFF << (PL_maxo & 0x07));
>                          ^
> ```
> 
> I've spotted this warning when compiling a fresh blead and couldn't remember seeing it earlier. There're no recent changes in ext/Opcode, so it's likely introduced by some unrelated change in types.
> 

Yes, I reported this in https://github.com/Perl/perl5/issues/17632.  But at that point it seemed as if the 'bug' had 'always been there; see https://github.com/Perl/perl5/issues/17632#issuecomment-598249646.  I had meant to open a separate ticket for that but forgot.

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> > This is a bug report for perl from [sergey.aleynikov@gmail.com](mailto:sergey.aleynikov@gmail.com),
> > generated with the help of perlbug 1.41 running under perl 5.31.10.
> > [Please describe your issue here]
> > ```
> > Opcode.xs: In function ‘XS_Opcode_invert_opset’:
> > Opcode.xs:363:25: warning: overflow in conversion from ‘int’ to ‘char’ changes value from ‘(int)*(bitmap + ((sizetype)my_cxt.x_opset_len + 18446744073709551615)) & -256’ to ‘0’ [-Woverflow]
> >   bitmap[opset_len-1] &= ~(0xFF << (PL_maxo & 0x07));
> >                          ^
> > ```
> > 
> > 
> > I've spotted this warning when compiling a fresh blead and couldn't remember seeing it earlier. There're no recent changes in ext/Opcode, so it's likely introduced by some unrelated change in types.
> 
> Yes, I reported this in #17632. But at that point it seemed as if the 'bug' had 'always been there; see [#17632 (comment)](https://github.com/Perl/perl5/issues/17632#issuecomment-598249646). I had meant to open a separate ticket for that but forgot.
> 
> Thank you very much.
> Jim Keenan

Per bisection, the `-Wformat` warning first appeared in this commit:
```
commit 02b85d3dab092d678cfc958a2dc252405333ed25
Author:     Zefram <zefram@fysh.org>
AuthorDate: Wed Feb 5 07:43:14 2020 +0000
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Thu Mar 12 22:34:26 2020 -0600

    chained comparisons
```
I have verified this using gcc version 7 on both Linux (unthreaded) and FreeBSD (threaded).
(Linux: gcc 7.5.0; FreeBSD-11: gcc 7.3.0)

```
$ parse-build-warnings 02b85d3dab.linux.gcc7.unthreaded.make.output.txt.gz 
File:  02b85d3dab.linux.gcc7.unthreaded.make.output.txt.gz

[
  {
    char   => 25,
    group  => "Woverflow",
    line   => 363,
    source => "Opcode.xs",
    text   => "overflow in implicit constant conversion",
  },
]
```

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> overflow in implicit constant conversion

Zefram has supplied a patch on list.  This is available for smoke-testing in this branch:
smoke-me/jkeenan/zefram/ghi-17664-opcode

**Comment by jkeenan** [comment]
> > overflow in implicit constant conversion
> 
> Zefram has supplied a patch on list. This is available for smoke-testing in this branch:
> smoke-me/jkeenan/zefram/ghi-17664-opcode

Following code review by @tonycoz, I merged this branch into blead in commit 95b7e3a36a7719a93600fbdbe66b690717cdff2d.  Will monitor smoke-test reports for several days, then close ticket.

Thank you very much.
Jim Keenan
